### ISSUE TITLE
Possibility to have flat redux structure

### ISSUE THREAD

**Issue by nosovsh** [omnidan/redux-undo#179]
Currently `redux-undo` adds one more level of hierarchy to your redux state:
```
{
  past: [...pastStatesHere...],
  present: {...currentStateHere...},
  future: [...futureStatesHere...]
}
```
Maybe it's a good default but if you already have a big application and want to start using `redux-undo` in different places you will face a problem that you have to change paths to you state in all your `connect` functions from `path.to.my.state` to `path.to.my.state.present`. Which can cause a lot of bugs and anyway not a very pleasant thing to do. My idea was to add option to `undoable` reducer that will force it to use such state structure instead:
```
{
  ...currentStateHere...
  past: [...pastStatesHere...],
  future: [...futureStatesHere...]
}
```
Setting option:
```
combineReducers({
  counter: undoable(counter, {
     flat: true,
  })
})
```

A far as I can see it can be easily implemented. What do you think?

**Comment by davidroeca** [comment]
See #154 for some discussion on this. While it's (somewhat) easy to implement, name collisions are the main issue we want to avoid, so if you can think of a way to handle name collisions I'm all ears

**Comment by nosovsh** [comment]
hi @davidroeca , thanks for the quick answer. I see that I'm not the first facing this problem.
Naming collisions can be prevented in few steps:
1. You have to explicitly set `flat: true` in order to use flat structure. So it is advanced scenario and users who know what they are doing will use it. 
2. We can raise warnings when user's reducer returns key that is used by undo.
3. And we can hide everything in one key to lower probability of collision, like 
```
{
  ...currentStateHere...
  __reduxUndoHistory: {
    past: [...pastStatesHere...],
    future: [...futureStatesHere...]
  }
}
```

@ugo-buonadonna I see that you make a fork with flat structure. Did you face any problems caused by such structure?

**Comment by davidroeca** [comment]
These are the trade-offs that make me less inclined to adopt this approach:
1. This breaks support of alternative state data structures such as the Immutable Map.
2. Specifying `flat: true` changes the behavior of the entire library and makes comparisons between the present and other elements of the undo/redo history much more resource-intense (instead of comparing references, the library will have to recursively check at the very least every relevant key/value of the object, with the exception of `_reduxUndoHistory`)

If the main headache is the selective use of the library at sub reducers, maybe an alternative to this is to make it more practical to wrap the combined reducers at the top level with redux-undo by making it easier to specify which subsections of the tree you want to track in the undo history. That way any refactoring is minimized to top-level state access

**Comment by davidroeca** [comment]
@nosovsh for the time being, would the following sort of filtering helper be useful at all?

```js
// say this is included in the library or in your code somewhere
export function includeSubsection(path, accessFunc) {
  if (accessFunc == null || typeof accessFunc !== 'function') {
    accessFunc = (state, path) => {
      let currentObj
      for (const key of path) {
        if (typeof currentObj === 'undefined') {
          return undefined
        }
        else {
          currentObj = current[key]
        }
      }
      return currentObj
    }
  }
  return (action, state, prevHist) => (
    accessFunc(state) !== accessFunc(prevHist._latestUnfiltered)
  )
}
```

The use case would be the following:

```js
import undoable, { includeSubsection, combineFilters } from 'redux-undo'
// assume we only want to undo the following portions of the undoable state:
// {
//   'userState1': {...I want to undo this state...}
//   'nested': {state: {with: {user: {state: {...I want to undo this state...}}}}},
//   ...I want redux-undo to not update the past with remaining keys...
// }
const finalReducer = undoable(reducer, {
  filter: combineFilters(
    includeSubsection(['userState1']),
    includeSubsection(['nested', 'state', 'with', 'user', 'state'])
  )
})
```

In theory you could also pass a memoized selector as the second argument to `includeSubsection` to improve performance. This would let you merely wrap your top-level reducer with redux-undo and only change your state tree in one place.

**Comment by nosovsh** [comment]
Do I understand correctly that you suggest to put  this at top level app reducer? And inside filter put all paths I need to make undoable?
So the state of the whole app would be 
```
{
  present: {
     'userState1': {...I want to undo this state...}
     'nested': {state: {with: {user: {state: {...I want to undo this state...}}}}},
     ...I want redux-undo to not update the past with remaining keys...
  },
  past: ...
  future: ...
}
```
right?

**Comment by davidroeca** [comment]
Yes. Or the highest level it needs to be, wherever you plan to put it. That way there are fewer `present`s to worry about

**Comment by davidroeca** [comment]
Also updated the `orFilters` function to do what the name intends (previously used the `&&` operator which is basically `combineFilters`)

**Comment by nosovsh** [comment]
But that means that all paths in app in connect/selectors should be updated even which don't need undo :)

I come up to the following reducer-wrapper that I can use in my code which will preserve my old state structure and will add undo state. What do you thing?
```javascript
// myReducer.js

// original reducer
const myReducer = (state={}, action) => ...

// reducer wrapper with undoable state
export default (state={}, action) => {
  const newUndoableState = undoable(myReducer)(state.undoable, action);
  if (newUndoableState === state.undoable) {
    return state;
  } else {
    return {
      ...undoableState.present,
      undoable: newUndoableState,
    }
  }
}
```

**Comment by davidroeca** [comment]
Hmm `undoableState.present` should be `newUndoableState.present`. But more importantly, the undoable reducer actually has a state of its own (initialization), so it shouldn't be created more than once. I'd say if anything, do the following:

```
// myReducer.js

// original reducer
const myReducer = (state={}, action) => ...

// the undoable portion
const undoableReducer = undoable(myReducer)
// reducer wrapper with undoable state
export default (state={}, action) => {
  const newUndoableState = undoableReducer(state.undoable, action);
  if (newUndoableState === state.undoable) {
    return state;
  } else {
    return {
      ...newUndoableState.present,
      undoable: newUndoableState,
    }
  }
}
```

Let me know if that gives you the desired behavior--should merely copy the references of the values in `newUndoableState` a level up

**Comment by nosovsh** [comment]
Thanks for tips. I will try it in real project this week

**Comment by dbkaplun** [comment]
This can be solved simply:
```js
{
  ...currentState,
  [PAST_KEY]: [...],
  [FUTURE_KEY]: [...]
}
export const PAST_KEY = '@@reduxUndo/past';
export const FUTURE_KEY = '@@reduxUndo/future';
```

[Prior art](https://github.com/reactjs/redux/blob/3e114f8c0fd4461e2f642c2737d2fa8297484728/src/createStore.js#L11)

**Comment by omnidan** [comment]
@dbkaplun that solves one of the problems. however, there are still other tradeoffs, as mentioned by @davidroeca earlier here: https://github.com/omnidan/redux-undo/issues/179#issuecomment-316475124

if we can find a solution to these issues, it would be great if we could switch to a flat structure, as it would make it even easier to integrate redux-undo.

**Comment by vincerubinetti** [comment]
For anyone stumbling across this, here's a solution I came up with that's similar to @nosovsh 's before I found this thread.

```javascript
export const undoer = (reducer) => ({ past = [], future = [], ...present } = {}, action) => {
  const newReducer = undoable(reducer, { OPTIONS });
  const newState = newReducer({ past, present, future }, action);
  return {
    ...newState.present,
    past: newState.past || [],
    present: newState.future || []
  };
};
```

You wrap your main/root reducer with this `undoer` higher-order reducer. I haven't tested this thoroughly but it seems to work.

It takes the incoming state (first argument, before `action`), and destructures the `past` and `future` keys out of it, and takes the rest of the keys as the `present`. Then it uses this library to convert your main reducer to an undoable one. Then it runs the reducer with the `past`, `present`, and `future` keys that this library expects and gets the new state. Finally it transforms this new state back into the schema you want, `{ ...state, past, future }`.

**Comment by posita** [comment]
I took a different approach:

```tsx
import { StateWithHistory } from "redux-undo"

export const wrapSelectorWithoutHistory = <StateT, ReturnT>(
  selector: ((state: StateT) => ReturnT)
): ((state: StateWithHistory<StateT> | StateT) => ReturnT) => {
  return (state: StateWithHistory<StateT> | StateT): ReturnT => {
    return selector(stateWithoutHistory(state))
  }
}

export const stateWithoutHistory = <T,>(state: StateWithHistory<T> | T): T => {
  if (doesStateHaveHistory(state)) {
    return state.present
  } else {
    return state
  }
}

const doesStateHaveHistory = <T,>(state: StateWithHistory<T> | T): state is StateWithHistory<T> => {
  return (
    "past" in state
    && "present" in state
    && "future" in state
  )
}
```

You can write your selector as if the state isn't undoable, then wrap it as follows:

```tsx
interface FooState {
  foo: number;
}

const mySelector = wrapSelectorWithoutHistory((state: FooState): number => {
  return state.foo
})
```

Alternatively, if you're being handed state elsewhere, you can do:

```tsx
const myHandler = (state: StateWithHistory<FooState> | FooState): void => {
  const rawState = stateWithoutHistory(state)
  const foo = rawState.foo
}
```
